File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
class Solution {
    public int solution(String dartResult) {
        int answer = 0;
        
        int[] dart = new int[3];
        
        int n = 0;
        int idx = 0;
        String numStr = "";
        
        for (int i=0;i<dartResult.length();i++) {
            char c = dartResult.charAt(i);
            
            // 숫자일때
            if (c >= '0' && c <= '9') {
                numStr += String.valueOf(c);
            } 
            // 보너스일때
            else if (c == 'S' || c == 'D' || c == 'T') {
                n = Integer.parseInt(numStr);
                
                if (c == 'S') {
                    dart[idx++] = (int)Math.pow(n,1);
                }
                else if (c == 'D') {
                    dart[idx++] = (int)Math.pow(n,2);
                } else {
                    dart[idx++] = (int)Math.pow(n,3);
                }
                numStr = "";
            }
            // 보너스일때
            else {
                if (c == '*') {
                    dart[idx-1] *= 2;
                    if (idx - 2 >= 0) dart[idx-2] *= 2;
                } else{
                    dart[idx-1] *= -1;
                }
            }
            
            
        }
        answer = dart[0] + dart[1] + dart[2];
        
        
        
        
        
        
        return answer;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> v;
string s;

int solution(string dartResult) {
    int answer = 0;

    for (char c : dartResult) {
        if (isdigit(c)) {
            s += c;
        } else {
            if (!s.empty()) {
                v.push_back(stoi(s));
                s = "";
            }

            if (c == 'S') continue;
            if (c == 'D') {
                v[v.size() - 1] = v[v.size() - 1] * v[v.size() - 1];
            }
            else if (c == 'T') {
                v[v.size() - 1] = v[v.size() - 1] * v[v.size() - 1] * v[v.size() - 1];
            }
            else if (c == '*') {
                v[v.size() - 1] *= 2;

                if (v.size() > 1) {
                    v[v.size() - 2] *= 2;
                }
            }
            else if (c == '#') {
                v[v.size() - 1] *= -1;
            }

        }
    }

    for (int result : v) answer += result;


    return answer;
}